File: test_cases/sk.py
# This would be one of your translations, as a Python dictionary
cases = {
    "Jasno": ["title", "clear"],
    "Možnosť veľmi slabých zrážok": ["title", "possible-very-light-precipitation"],
    "Slabé zrážky": ["title", "very-light-precipitation"],
    "Možnosť slabých zrážok": ["title", "possible-light-precipitation"],
    "Slabé zrážky": ["title", "light-precipitation"],
    "Zrážky": ["title", "medium-precipitation"],
    "Silné zrážky": ["title", "heavy-precipitation"],
    "Možnosť mrholenia": ["title", "possible-very-light-rain"],
    "Mrholenie": ["title", "very-light-rain"],
    "Možnosť slabého dažďa": ["title", "possible-light-rain"],
    "Slabý dážď": ["title", "light-rain"],
    "Dážď": ["title", "medium-rain"],
    "Vydatný dážď": ["title", "heavy-rain"],
    "Možnosť slabého dažďa so snehom": ["title", "possible-very-light-sleet"],
    "Slabý dážď so snehom": ["title", "very-light-sleet"],
    "Možnosť slabého dažďa so snehom": ["title", "possible-light-sleet"],
    "Slabý dážď so snehom": ["title", "light-sleet"],
    "Dážď so snehom": ["title", "medium-sleet"],
    "Vydatný dážď so snehom": ["title", "heavy-sleet"],
    "Možnosť slabého sneženia": ["title", "possible-very-light-snow"],
    "Slabé sneženie": ["title", "very-light-snow"],
    "Možnosť slabého sneženia": ["title", "possible-light-snow"],
    "Slabé sneženie": ["title", "light-snow"],
    "Sneženie": ["title", "medium-snow"],
    "Vydatné sneženie": ["title", "heavy-snow"],
    "Veterno": ["title", "medium-wind"],
    "Silný vietor": ["title", "heavy-wind"],
    "Hmlisto": ["title", "fog"],
    "Prevažne jasno": ["title", "very-light-clouds"],
    "Čiastočne zamračené": ["title", "light-clouds"],
    "Prevažne zamračené": ["title", "medium-clouds"],
    "Zamračené": ["title", "heavy-clouds"],
    "Nízka vlhkosť a slabý vietor": ["title", ["and", "low-humidity", "light-wind"]],
    "Mrholenie a silný vietor": ["title", ["and", "very-light-rain", "heavy-wind"]],
    "Vysoká vlhkosť a čiastočne zamračené": [
        "title",
        ["and", "high-humidity", "light-clouds"],
    ],
    "Jasno hodinu.": ["sentence", ["for-hour", "clear"]],
    "Slabé sneženie o 35 min.": [
        "sentence",
        ["starting-in", "very-light-snow", ["minutes", 35]],
    ],
    "Slabý dážď skončí o 15 min.": [
        "sentence",
        ["stopping-in", "light-rain", ["minutes", 15]],
    ],
    "Vydatný dážď so snehom o 20 min., skončí o 30 min. neskôr.": [
        "sentence",
        [
            "starting-then-stopping-later",
            "heavy-sleet",
            ["minutes", 20],
            ["minutes", 30],
        ],
    ],
    "Dážď skončí o 25 min. a začne znovu o 8 min. neskôr.": [
        "sentence",
        [
            "stopping-then-starting-later",
            "medium-rain",
            ["minutes", 25],
            ["minutes", 8],
        ],
    ],
    "Počas dňa prevažne zamračené.": ["sentence", ["for-day", "medium-clouds"]],
    "Od rána slabý dážď so snehom.": [
        "sentence",
        ["starting", "very-light-sleet", "morning"],
    ],
    "Veterno až do dnešnej noci.": [
        "sentence",
        ["until", "medium-wind", "today-night"],
    ],
    "Silné zrážky až do popoludnia.": [
        "sentence",
        ["until", "heavy-precipitation", "afternoon"],
    ],
    "Popoludní slabý vietor.": ["sentence", ["during", "light-wind", "afternoon"]],
    "Dnes neskoro večer a zajtra ráno sneženie.": [
        "sentence",
        ["during", "medium-snow", ["and", "later-today-evening", "tomorrow-morning"]],
    ],
    "Vydatný dážď až do dnešného dopoludnia, ktorý začne znovu dnes večer.": [
        "sentence",
        ["until-starting-again", "heavy-rain", "later-today-morning", "today-evening"],
    ],
    "Od dnešnej neskorej noci zamračené a zajtra popoludní vydatné sneženie.": [
        "sentence",
        ["starting-continuing-until", "heavy-clouds", "evening", "night"],
    ],
    "Od popoludnia čiastočne zamračené a pretrvá až do večera.": [
        "sentence",
        ["starting-continuing-until", "light-clouds", "afternoon", "evening"],
    ],
    "Dnes podvečer slabý dážď so snehom a zajtra ráno hmlisto.": [
        "sentence",
        [
            "and",
            ["during", "light-sleet", "later-today-afternoon"],
            ["during", "fog", "tomorrow-morning"],
        ],
    ],
    "Od dnešného rána silný vietor, ktorý pretrvá až do dnešného popoludnia a zajtra ráno dážď so snehom.": [
        "sentence",
        [
            "and",
            [
                "starting-continuing-until",
                "heavy-wind",
                "today-morning",
                "today-afternoon",
            ],
            ["during", "medium-sleet", "tomorrow-morning"],
        ],
    ],
    "Od dnešnej neskorej noci zamračené a zajtra popoludní vydatné sneženie.": [
        "sentence",
        [
            "and",
            ["starting", "heavy-clouds", "later-today-night"],
            ["during", "heavy-snow", "tomorrow-afternoon"],
        ],
    ],
    "Dnes v noci nízka vlhkosť a od zajtrajšieho večera slabé zrážky, ktoré pretrvajú až do zajtrajšej noci.": [
        "sentence",
        [
            "and",
            ["during", "low-humidity", "today-night"],
            [
                "starting-continuing-until",
                "light-precipitation",
                "tomorrow-evening",
                "tomorrow-night",
            ],
        ],
    ],
    "V noci sneženie (5 in).": [
        "sentence",
        ["during", ["parenthetical", "medium-snow", ["inches", 5]], "night"],
    ],
    "Dnes dopoludnia slabé sneženie (2 cm).": [
        "sentence",
        [
            "during",
            ["parenthetical", "light-snow", ["centimeters", 2]],
            "later-today-morning",
        ],
    ],
    "Počas dňa vydatné sneženie (8-12 in).": [
        "sentence",
        ["for-day", ["parenthetical", "heavy-snow", ["inches", ["range", 8, 12]]]],
    ],
    "Popoludní sneženie (menej ako 1 cm).": [
        "sentence",
        [
            "during",
            ["parenthetical", "medium-snow", ["less-than", ["centimeters", 1]]],
            "afternoon",
        ],
    ],
    "Bez zrážok počas týždňa, zajtra s teplotným maximom 85°F.": [
        "sentence",
        [
            "with",
            ["for-week", "no-precipitation"],
            ["temperatures-peaking", ["fahrenheit", 85], "tomorrow"],
        ],
    ],
    "Zmiešané zrážky cez víkend, vo štvrtok s teplotami stúpajúcimi k 32°C.": [
        "sentence",
        [
            "with",
            ["over-weekend", "mixed-precipitation"],
            ["temperatures-rising", ["celsius", 32], "thursday"],
        ],
    ],
    "V pondelok mrholenie, v piatok s teplotným minimom 15°F.": [
        "sentence",
        [
            "with",
            ["during", "very-light-rain", "monday"],
            ["temperatures-valleying", ["fahrenheit", 15], "friday"],
        ],
    ],
    "V utorok a budúcu stredu slabé sneženie, v nedeľu s teplotami klesajúcimi k 0°C.": [
        "sentence",
        [
            "with",
            ["during", "light-snow", ["and", "tuesday", "next-wednesday"]],
            ["temperatures-falling", ["celsius", 0], "sunday"],
        ],
    ],
    "Od dnes do soboty zrážky, v pondelok s teplotným maximom 100°F.": [
        "sentence",
        [
            "with",
            ["during", "medium-precipitation", ["through", "today", "saturday"]],
            ["temperatures-peaking", ["fahrenheit", 100], "monday"],
        ],
    ],
    "Počas dňa zmiešané zrážky (1-3 in snehu).": [
        "sentence",
        [
            "for-day",
            ["parenthetical", "mixed-precipitation", ["inches", ["range", 1, 3]]],
        ],
    ],
    "Vydatné sneženie (1-3 in)": [
        "title",
        ["parenthetical", "heavy-snow", ["inches", ["range", 1, 3]]],
    ],
    "Vydatné sneženie (3-5 cm)": [
        "title",
        ["parenthetical", "heavy-snow", ["centimeters", ["range", 3, 5]]],
    ],
    "Možnosť búrok": ["title", "possible-thunderstorm"],
    "Vydatný dážď a búrky": ["title", ["and", "heavy-rain", "thunderstorm"]],
    "Búrky až do budúceho pondelka.": [
        "sentence",
        ["until", "thunderstorm", "next-monday"],
    ],
    "Predpoveď na ďalšiu hodinu je dočasne nedostupná, pretože všetky radarové stanice v okolí sú v režime offline.": [
        "sentence",
        ["next-hour-forecast-status", "temporarily-unavailable", "station-offline"],
    ],
    "Predpoveď na ďalšiu hodinu je čiastočne nedostupná, pretože vznikli medzery v pokrytí radarovými stanicami v okolí.": [
        "sentence",
        ["next-hour-forecast-status", "partially-unavailable", "station-incomplete"],
    ],
    "Predpoveď na ďalšiu hodinu je nedostupná, pretože všetky radarové stanice v okolí sú v režime offline.": [
        "sentence",
        ["next-hour-forecast-status", "unavailable", "station-offline"],
    ],
}

File: lib/lang/sk.js


function remove_prefix_and_use_genitive(a) {
  switch(a) {
    case "dnes ráno":
      return "dnešného rána";
    case "dnes dopoludnia":
      return "dnešného dopoludnia";
    case "dnes popoludní":
      return "dnešného popoludnia";
    case "dnes podvečer":
      return "dnešného podvečera";
    case "dnes večer":
      return "dnešného večera";
    case "dnes neskoro večer":
      return "dnešného neskorého večera";
    case "dnes v noci":
      return "dnešnej noci";
    case "dnes neskoro v noci":
      return "dnešnej neskorej noci";
    case "zajtra ráno":
      return "zajtrajšieho rána";
    case "zajtra popoludní":
      return "zajtrajšieho popoludnia";
    case "zajtra večer":
      return "zajtrajšieho večera";
    case "zajtra v noci":
      return "zajtrajšej noci";
    case "ráno":
      return "rána";
    case "popoludní":
      return "popoludnia";
    case "večer":
      return "večera";
    case "v noci":
      return "noci";
    case "v pondelok":
      return "pondelka";
    case "v utorok":
      return "utorka";
    case "v stredu":
      return "stredy";
    case "vo štvrtok":
      return "štvrtka";
    case "v piatok":
      return "piatka";
    case "v sobotu":
      return "soboty";
    case "v nedeľu":
      return "nedele";
    case "budúci pondelok":
      return "budúceho pondelka";
    case "budúci utorok":
      return "budúceho utorka";
    case "budúca streda":
      return "budúcej stredy";
    case "budúci štvrtok":
      return "budúceho štvrtka";
    case "budúci piatok":
      return "budúceho piatka";
    case "budúca sobota":
      return "budúcej soboty";
    case "budúca nedeľa":
      return "budúcej nedele";
    default:
      return a;
  }
}

function join_with_shared_prefix(a, b, joiner) {
  let i = 0;

  while(i !== a.length &&
        i !== b.length &&
        a.charCodeAt(i) === b.charCodeAt(i))
    ++i;

  while(i && a.charCodeAt(i - 1) !== 32)
    --i;

  return a + joiner + b.slice(i);
}

module.exports = {
  "clear": "jasno",
  "no-precipitation": "bez zrážok",
  "mixed-precipitation": "zmiešané zrážky",
  "possible-very-light-precipitation": "možnosť veľmi slabých zrážok",
  "very-light-precipitation": "veľmi slabé zrážky",
  "possible-light-precipitation": "možnosť slabých zrážok",
  "light-precipitation": "slabé zrážky",
  "medium-precipitation": "zrážky",
  "heavy-precipitation": "silné zrážky",
  "possible-very-light-rain": "možnosť mrholenia",
  "very-light-rain": "mrholenie",
  "possible-light-rain": "možnosť slabého dažďa",
  "light-rain": "slabý dážď",
  "medium-rain": "dážď",
  "heavy-rain": "vydatný dážď",
  "possible-very-light-sleet": "možnosť slabého dažďa so snehom",
  "very-light-sleet": "slabý dážď so snehom",
  "possible-light-sleet": "možnosť slabého dažďa so snehom",
  "light-sleet": "slabý dážď so snehom",
  "medium-sleet": "dážď so snehom",
  "heavy-sleet": "vydatný dážď so snehom",
  "possible-very-light-snow": "možnosť slabého sneženia",
  "very-light-snow": "slabé sneženie",
  "possible-light-snow": "možnosť slabého sneženia",
  "light-snow": "slabé sneženie",
  "medium-snow": "sneženie",
  "heavy-snow": "vydatné sneženie",
  "possible-thunderstorm": "možnosť búrok",
  "thunderstorm": "búrky",
  "possible-medium-precipitation": "possible precipitation",
  "possible-heavy-precipitation": "possible heavy precipitation",
  "possible-medium-rain": "possible rain",
  "possible-heavy-rain": "possible heavy rain",
  "possible-medium-sleet": "possible sleet",
  "possible-heavy-sleet": "possible heavy sleet",
  "possible-medium-snow": "possible snow",
  "possible-heavy-snow": "possible heavy snow",
  "possible-freezing-drizzle": "possible freezing drizzle",
  "freezing-drizzle": "freezing drizzle",
  "possible-light-freezing-rain": "possible light freezing rain",
  "light-freezing-rain": "light freezing rain",
  "possible-freezing-rain": "possible freezing rain",
  "freezing-rain": "freezing rain",
  "possible-heavy-freezing-rain": "possible heavy freezing rain",
  "heavy-freezing-rain": "heavy freezing rain",
  "possible-hail": "possible hail",
  "hail": "hail",
  "light-wind": "slabý vietor",
  "medium-wind": "veterno",
  "heavy-wind": "silný vietor",
  "low-humidity": "nízka vlhkosť",
  "high-humidity": "vysoká vlhkosť",
  "fog": "hmlisto",
  "very-light-clouds": "prevažne jasno",
  "light-clouds": "čiastočne zamračené",
  "medium-clouds": "prevažne zamračené",
  "heavy-clouds": "zamračené",
  "today-morning": "dnes ráno",
  "later-today-morning": "dnes dopoludnia",
  "today-afternoon": "dnes popoludní",
  "later-today-afternoon": "dnes podvečer",
  "today-evening": "dnes večer",
  "later-today-evening": "dnes neskoro večer",
  "today-night": "dnes v noci",
  "later-today-night": "dnes neskoro v noci",
  "tomorrow-morning": "zajtra ráno",
  "tomorrow-afternoon": "zajtra popoludní",
  "tomorrow-evening": "zajtra večer",
  "tomorrow-night": "zajtra v noci",
  "morning": "ráno",
  "afternoon": "popoludní",
  "evening": "večer",
  "night": "v noci",
  "today": "dnes",
  "tomorrow": "zajtra",
  "sunday": "v nedeľu",
  "monday": "v pondelok",
  "tuesday": "v utorok",
  "wednesday": "v stredu",
  "thursday": "vo štvrtok",
  "friday": "v piatok",
  "saturday": "v sobotu",
  "next-sunday": "budúcu nedeľu",
  "next-monday": "budúci pondelok",
  "next-tuesday": "budúci utorok",
  "next-wednesday": "budúcu stredu",
  "next-thursday": "budúci štvrtok",
  "next-friday": "budúci piatok",
  "next-saturday": "budúcu sobotu",
  "minutes": "$1 min.",
  "fahrenheit": "$1°F",
  "celsius": "$1°C",
  "inches": "$1 in",
  "centimeters": "$1 cm",
  "less-than": "menej ako $1",
  "and": function(a, b) {
    return join_with_shared_prefix(
      a,
      b,
      " a ",
    );
  },
  "through": function(a, b) {
    return "od " + remove_prefix_and_use_genitive(a) + " do " + remove_prefix_and_use_genitive(b);
  },
  "with": "$1, $2",
  "range": "$1-$2",
  "parenthetical": function(a, b) {
    return a + " (" + b + (a === "zmiešané zrážky" ? " snehu)" : ")");
  },
  "for-hour": "$1 hodinu",
  "starting-in": "$1 o $2",
  "stopping-in": "$1 skončí o $2",
  "starting-then-stopping-later": "$1 o $2, skončí o $3 neskôr",
  "stopping-then-starting-later": "$1 skončí o $2 a začne znovu o $3 neskôr",
  "for-day": "Počas dňa $1",
  "starting": function(a, b) {
    return "od " + remove_prefix_and_use_genitive(b) + " " + a;
  },
  "until": function(condition, period) {
    return condition + " až do " + remove_prefix_and_use_genitive(period);
  },
  "until-starting-again": function(condition, a, b) {
    let starting = "";
    if(condition.endsWith("zrážky"))
      starting = ", ktoré začnú znovu ";
    else if(condition.endsWith("dážď") || condition.endsWith("dážď so snehom") || condition.endsWith("vietor"))
      starting = ", ktorý začne znovu ";
    else if(condition.endsWith("sneženie") || condition.endsWith("mrholenie"))
      starting = ", ktoré začne znovu ";
    else if(condition.endsWith("vlhkosť"))
      starting = ", ktorá začne znovu ";
    else if(condition.endsWith("zamračené") || condition.endsWith("hmlisto"))
      starting = "a začne znovu ";

    return condition + " až do " + remove_prefix_and_use_genitive(a) + starting + b;
  },
  "starting-continuing-until": function(condition, a, b) {
    let continuing = "";
    if(condition.endsWith("zrážky"))
      continuing = ", ktoré pretrvajú až do ";
    else if(condition.endsWith("dážď") || condition.endsWith("dážď so snehom") || condition.endsWith("vietor"))
      continuing = ", ktorý pretrvá až do ";
    else if(condition.endsWith("sneženie") || condition.endsWith("mrholenie"))
      continuing = ", ktoré pretrvá až do ";
    else if(condition.endsWith("vlhkosť"))
      continuing = ", ktorá pretrvá až do ";
    else if(condition.endsWith("zamračené") || condition.endsWith("hmlisto"))
      continuing = " a pretrvá až do ";
    return "od " + remove_prefix_and_use_genitive(a) + " " + condition + continuing + remove_prefix_and_use_genitive(b);
  },
  "during": "$2 $1",
  "for-week": "$1 počas týždňa",
  "over-weekend": "$1 cez víkend",
  "temperatures-peaking": "$2 s teplotným maximom $1",
  "temperatures-rising": "$2 s teplotami stúpajúcimi k $1",
  "temperatures-valleying": "$2 s teplotným minimom $1",
  "temperatures-falling": "$2 s teplotami klesajúcimi k $1",
  /* Capitalize the first letter of first word. */
  "title": function(str) {
    return str.charAt(0).toUpperCase() + str.slice(1);
  },
  /* Capitalize the first word of the sentence and end with a period. */
  "sentence": function(str) {
    /* Capitalize. */
    str = str.charAt(0).toUpperCase() + str.slice(1);

    /* Add a period if there isn't already one. */
    if(str.charAt(str.length - 1) !== ".")
      str += ".";

    return str;
  },
  "next-hour-forecast-status": "Predpoveď na ďalšiu hodinu je $1, pretože $2.",
  "unavailable": "nedostupná",
  "temporarily-unavailable": "dočasne nedostupná",
  "partially-unavailable": "čiastočne nedostupná",
  "station-offline": "všetky radarové stanice v okolí sú v režime offline",
  "station-incomplete": "vznikli medzery v pokrytí radarovými stanicami v okolí",
};
